player/lru_cache.go

```diff
@@ -0,0 +1,149 @@
+package player
+
+import (
+	"fmt"
+	"os"
+	"path/filepath"
+	"time"
+
+	lru "github.com/hashicorp/golang-lru"
+	"github.com/lbryio/lbry.go/v2/stream"
+)
+
+// LRUCacheOpts contains options for a cache. Size is max size in bytes.
+type LRUCacheOpts struct {
+	Path          string
+	Size          uint64
+	SweepInterval time.Duration
+}
+
+type lruCache struct {
+	storage *fsStorage
+	lru     *lru.Cache
+}
+
+// InitLRUCache initializes a LRU cache for chunks.
+func InitLRUCache(opts *LRUCacheOpts) (ChunkCache, error) {
+	storage, err := initFSStorage(opts.Path)
+	if err != nil {
+		return nil, err
+	}
+
+	if opts.Size == 0 {
+		opts.Size = defaultMaxCacheSize
+	}
+
+	onEvicted := func(key interface{}, value interface{}) {
+		storage.remove(value)
+	}
+
+	lru, err := lru.NewWithEvict(int(opts.Size/ChunkSize), onEvicted)
+	if err != nil {
+		return nil, err
+	}
+
+	c := &lruCache{storage, lru}
+
+	Logger.Infof("LRU cache of %vGB initialized at %v", opts.Size/1024/1024/1024, opts.Path)
+
+	go func() {
+		Logger.Infoln("restoring cache in memory...")
+		err := c.reloadCache()
+		if err != nil {
+			Logger.Errorf("failed to restore cache in memory: %s", err.Error())
+		}
+	}()
+
+	return c, nil
+}
+
+func (c *lruCache) reloadCache() error {
+	err := filepath.Walk(c.storage.path, func(path string, info os.FileInfo, err error) error {
+		if c.storage.path == path {
+			return nil
+		}
+		if err != nil {
+			return err
+		}
+		if info.IsDir() {
+			return fmt.Errorf("subfolder %v found inside cache folder", path)
+		}
+		if len(info.Name()) != stream.BlobHashHexLength {
+			return fmt.Errorf("non-cache file found at path %v", path)
+		}
+		c.lru.Add(info.Name(), info.Name())
+		return nil
+	})
+	return err
+}
+
+func (c *lruCache) Set(hash string, body []byte) (ReadableChunk, error) {
+	var numWritten int
+	Logger.Debugf("attempting to cache chunk %v", hash)
+	chunkPath := c.storage.getPath(hash)
+
+	f, err := os.OpenFile(chunkPath, os.O_WRONLY|os.O_CREATE|os.O_EXCL, 0600)
+	if os.IsExist(err) {
+		MetCacheErrorCount.Inc()
```

Comment:
where is this declared? is it thread safe?

Reply:
It is prometheus metrics, completely safe.